dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java

```diff
@@ -700,47 +637,64 @@ else if ( STATIC_DIMS.contains( dimension ) )
+            TrackedEntityAttributeDimension::getUid );
 
-                return new BaseDimensionalObject( dimension, DimensionType.PROGRAM_DATA_ELEMENT, null,
-                    tedd.getDisplayName(), tedd.getLegendSet(), tedd.getProgramStage(), tedd.getFilter() );
-            }
+        if ( attributes.containsKey( dimension ) )
+        {
+            final TrackedEntityAttributeDimension tead = attributes.get( dimension );
+            final ValueType valueType = tead.getAttribute() != null ? tead.getAttribute().getValueType() : null;
+            final OptionSet optionSet = tead.getAttribute() != null ? tead.getAttribute().getOptionSet() : null;
 
-            // Tracked entity program indicator
+            return new BaseDimensionalObject( dimension, DimensionType.PROGRAM_ATTRIBUTE, null,
+                tead.getDisplayName(), tead.getLegendSet(), null, tead.getFilter(), valueType, optionSet );
+        }
 
-            Map<String, TrackedEntityProgramIndicatorDimension> programIndicators = Maps
-                .uniqueIndex( programIndicatorDimensions, TrackedEntityProgramIndicatorDimension::getUid );
+        // Tracked entity data element
 
-            if ( programIndicators.containsKey( dimension ) )
-            {
-                TrackedEntityProgramIndicatorDimension teid = programIndicators.get( dimension );
+        final Map<String, TrackedEntityDataElementDimension> dataElements = Maps.uniqueIndex( dataElementDimensions,
+            TrackedEntityDataElementDimension::getUid );
 
-                return new BaseDimensionalObject( dimension, DimensionType.PROGRAM_INDICATOR, null,
-                    teid.getDisplayName(), teid.getLegendSet(), null, teid.getFilter() );
-            }
+        if ( dataElements.containsKey( dimension ) )
+        {
+            final TrackedEntityDataElementDimension tedd = dataElements.get( dimension );
+            final ValueType valueType = tedd.getDataElement() != null ? tedd.getDataElement().getValueType() : null;
+            final OptionSet optionSet = tedd.getDataElement() != null ? tedd.getDataElement().getOptionSet() : null;
+
+            return new BaseDimensionalObject( dimension, DimensionType.PROGRAM_DATA_ELEMENT, null,
+                tedd.getDisplayName(), tedd.getLegendSet(), tedd.getProgramStage(),
+                tedd.getFilter(), valueType, optionSet );
         }
 
-        throw new IllegalArgumentException( "Not a valid dimension: " + dimension );
+        // Tracked entity program indicator
+
+        final Map<String, TrackedEntityProgramIndicatorDimension> programIndicators = Maps
+            .uniqueIndex( programIndicatorDimensions, TrackedEntityProgramIndicatorDimension::getUid );
+
+        if ( programIndicators.containsKey( dimension ) )
+        {
+            final TrackedEntityProgramIndicatorDimension teid = programIndicators.get( dimension );
```

Comment:
teid still can be null, can you please add the test to avoid null ptr exception?

Reply:
And also here